src/nvim/api/autocmd.c

```diff
@@ -849,16 +844,16 @@ void nvim_exec_autocmds(Object event, Dict(exec_autocmds) *opts, Error *err)
 
   modeline = api_object_to_bool(opts->modeline, "modeline", true, err);
 
-  if (set_pattern && set_buf) {
-    api_set_error(err, kErrorTypeValidation, "must not set 'buffer' and 'pattern'");
-    goto cleanup;
-  }
-
   bool did_aucmd = false;
   FOREACH_ITEM(event_array, event_str, {
     GET_ONE_EVENT(event_nr, event_str, cleanup)
 
-    did_aucmd |= apply_autocmds_group(event_nr, pattern, NULL, true, au_group, buf, NULL, data);
+    for (size_t i = 0; i < patterns.size; i++) {
+      Object pat = patterns.items[i];
+      char *fname = opts->buffer.type == kObjectTypeNil ? pat.data.string.data : NULL;
+      did_aucmd |=
+        apply_autocmds_group(event_nr, fname, NULL, true, au_group, buf, NULL, data);
+    }
```

Comment:
Could also use `FOREACH_ITEM` here.

```suggestion
    FOREACH_ITEM(patterns, pat, {
      char *fname = opts->buffer.type == kObjectTypeNil ? pat.data.string.data : NULL;
      did_aucmd |=
        apply_autocmds_group(event_nr, fname, NULL, true, au_group, buf, NULL, data);
    }
```

Reply:
Good idea; will also backport that pattern to `nvim_create_autocommands` (where I stole this from).

There's quite a bit of refactoring opportunities on this table for people so inclined ;)